init/plugins/libfido2_plugin.go

```diff
@@ -0,0 +1,304 @@
+	// if missing parameter.
+	C.FIDO_ERR_MISSING_PARAMETER: fmt.Errorf("missing parameter"), // 0x14)
+	// if not allowed.
+	C.FIDO_ERR_NOT_ALLOWED: fmt.Errorf("not allowed"),
+	// if action timed out.
+	C.FIDO_ERR_ACTION_TIMEOUT: fmt.Errorf("action timed out"),
+	// if PIN is not set and is required for command.
+	C.FIDO_ERR_PIN_NOT_SET: fmt.Errorf("pin not set"),
+	// if credential is invalid.
+	C.FIDO_ERR_INVALID_CREDENTIAL: fmt.Errorf("invalid credential"),
+	// if option is unsupported.
+	C.FIDO_ERR_UNSUPPORTED_OPTION: fmt.Errorf("unsupported option"),
+	// if pin is wrong.
+	C.FIDO_ERR_PIN_INVALID: fmt.Errorf("pin invalid"),
+	// rx not CBOR.
+	C.FIDO_ERR_RX_NOT_CBOR: fmt.Errorf("rx not CBOR"),
+	// internal error.
+	C.FIDO_ERR_INTERNAL: fmt.Errorf("internal error"),
+	// if PIN policy violation.
+	C.FIDO_ERR_PIN_POLICY_VIOLATION: fmt.Errorf("pin policy violation"),
+	// if no credentials.
+	C.FIDO_ERR_NO_CREDENTIALS: fmt.Errorf("no credentials"),
+	// if too many PIN failures.
+	C.FIDO_ERR_PIN_AUTH_BLOCKED: fmt.Errorf("pin auth blocked"),
+	// if PIN is required.
+	C.FIDO_ERR_PIN_REQUIRED: fmt.Errorf("pin required"),
+	// if user presence is required.
+	C.FIDO_ERR_UP_REQUIRED: fmt.Errorf("up required"),
+	// if receiving invalid CBOR.
+	C.FIDO_ERR_RX_INVALID_CBOR: fmt.Errorf("rx invalid cbor"),
+	// if operation denied.
+	C.FIDO_ERR_OPERATION_DENIED: fmt.Errorf("operation denied"),
+	// if action was cancelled.
+	C.FIDO_ERR_KEEPALIVE_CANCEL: fmt.Errorf("keep alive cancel"),
+	// if option is invalid.
+	C.FIDO_ERR_INVALID_OPTION: fmt.Errorf("invalid option"),
+	C.FIDO_ERR_ERR_OTHER:      fmt.Errorf("other error"),
+}
+
+/*
+	initiliaze the library
+
+The fido_init() function initialises the libfido2 library. Its invocation must precede that of any other libfido2 function in the context of the executing thread.
+If FIDO_DEBUG is set in flags, then debug output will be emitted by libfido2 on stderr. Alternatively, the FIDO_DEBUG environment variable may be set.
+If FIDO_DISABLE_U2F_FALLBACK is set in flags, then libfido2 will not fallback to U2F in fido_dev_open(3) if a device claims to support FIDO2 but fails to respond to a CTAP 2.0 greeting.
+
+- https://developers.yubico.com/libfido2/Manuals/fido_init.html
+*/
+func init() {
+	C.fido_init(0) // initiliaze the library without debugging
+}
+
+func newFido2Device(path string) *Device {
+	return &Device{
+		path: path,
+	}
+}
+
+func (d *Device) openFido2Device() (*C.fido_dev_t, error) {
+	dev := C.fido_dev_new()
+	if cErr := C.fido_dev_open(dev, C.CString(d.path)); cErr != C.FIDO_OK {
+		return nil, fmt.Errorf("failed to open hidraw device: %w", libfido2Errors[cErr])
+	}
+	d.dev = dev
+
+	return dev, nil
+}
+
+func (d *Device) closeFido2Device(dev *C.fido_dev_t) {
+	d.Lock()
+	d.dev = nil
+	d.Unlock()
+	if cErr := C.fido_dev_close(dev); cErr != C.FIDO_OK {
+		fmt.Println("failed to close hidraw device: %w", libfido2Errors[cErr])
```

Comment:
This is useless and would like to use `info()` here, but since we're inside the plugins directory Go is unable to find its reference. Is there a way to reference it besides moving the file where `logging.go` is?

Reply:
I am not good with `plugin` subsystem TBH and don't really know if there is a good way to do it.

assuming that it is a rare error it is ok to leave `fmt.Println` like this for now.